docs/proposals/0000-support-interchangeable-build-and-run-implementations.md

```diff
@@ -0,0 +1,248 @@
+- Applying conventions to pod spec, such as labels, annotations, volume mounts, etc.
+
+We could accomplish this flexibility with a smaller change: splitting out the current reconcilers for `CFBuild` and
+`CFProcess` into their own repositories, so they can be replaced with alternatives. The issue with this change is that
+the existing reconcilers use information from many resources and an alternative implementation would have to duplicate
+this complex logic. For example, to create the LRPs that currently generate the run step's stateful set, the `CFProcess`
+reconciler uses `CFProcess`, `CFApp`, `CFPackage`, `CFServiceBinding`, and `Secret` resources.
+
+The new `BuildWorkload` and `RunWorkload` solve this problem by placing all necessary information on a single resource.
+As a consequence, the build and run controllers can focus on implementing their respective flows without having to fetch
+and correlate multiple resources. These resources also provide a standard way to report status back to
+the `korifi-controllers` without directly modifying Korifi domain resources.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Here are the various resources and components involved:
+
+- `BuildWorkload` - a new Custom Resource Definition, with fields that contain all necessary information to execute the
+  build step for a korifi workload. This resource contains information from `CFBuild`, `CFPackage`, `CFServiceBinding`,
+  and the App's `Secret` resources. This CRD may exist in a new, separate API Group to indicate its status as an
+  external interface.
+
+- `RunWorkload` - a new Custom Resource Definition, with fields that contain all necessary information to execute the
+  run step for a korifi workload. This will be similar to the existing Eirini `LRP` CRD, but with some improvements to
+  simplify construction of a pod spec template. This resource contains information from `CFApp`, `CFProcess`,
+  `CFBuild/CFDroplet`, and the App's `Secret` resources. This CRD may exist in a new, separate API Group to indicate its
+  status as an external interface. We may even want to reclaim the `workloads.cloudfoundry.org` group for this purpose
+  and choose another for the existing resources.
+
+- build workload reconciler - a separate component from `korifi-controllers` that reconciles `BuildWorkload` resources
+  to execute the build workflow. Eventually multiple implementations of this reconciler will exist, each implementing
+  the build workflow in a different way to meet operator needs. These CRDs may exist in a new, separate API Group to
+  indicate their status as an external interface. We may even want to reclaim the `workloads.cloudfoundry.org` group for
+  this purpose and choose another for the existing resources.
+
+- run workload reconciler - similar to the build workload reconciler, but for `RunWorkload` resources.
+
+- the `CFBuild` reconciler: a component of `korifi-controllers` that creates `BuildWorkload` resources and monitors
```

Comment:
Here's an idea -- I'm not sure where the best place to put this is, but will just drop it here.

What if we had forego some of the V3 CF behavior around having multiple concurrent builds and instead behaved more like V2 CF and only allowed a single build to be in flight (I think in V2 subsequent builds would either fail or supersede the ongoing one). This would allow us to have a single `BuildWorkload` per app which might simplify things and the `CFBuild` controller could just update this with the new package info, etc.

This would have some benefits:
1. In our current implementation there would be a single `kpack Image` per app, so we could leverage some of the caching behavior it has. We just push toy apps right now like the node app or staticfile app, but real world users with large Java apps (or node apps with a lot of dependencies) would appreciate this.
2. I think this would more naturally support the auto-rebuild functionality of kpack when a stack gets updated. One of the things we're missing in Korifi is automatic rollout of stack updates (like we have in CF for VMs when Diego cells get rolled and apps evacuate) and this would simplify our design and support of this.
3. I think it might also potentially make it easier to extend for implementing some of the alternative build systems I can think of.